client/my-sites/controller.js

```diff
@@ -109,6 +112,43 @@ function renderNoVisibleSites( context ) {
 	);
 }
 
+function isPathAllowedForDomainOnlySite( pathname ) {
+	const urlPrefixesWhiteListForDomainOnlySite = [
+		'/domains/manage/',
+		'/checkout/',
+	];
+
+	return urlPrefixesWhiteListForDomainOnlySite.some( path => startsWith( pathname, path ) );
+}
+
+function onSelectedSiteAvailable( context ) {
+	const selectedSite = sites.getSelectedSite();
+	siteStatsStickyTabActions.saveFilterAndSlug( false, selectedSite.slug );
+	context.store.dispatch( receiveSite( selectedSite ) );
+	context.store.dispatch( setSelectedSiteId( selectedSite.ID ) );
+
+	const state = context.store.getState();
+
+	if ( isDomainOnlySite( state, selectedSite.ID ) &&
+		! isPathAllowedForDomainOnlySite( context.pathname ) ) {
+		page.redirect( domainManagementList( selectedSite.slug ) );
+		return false;
+	}
+
+	// Update recent sites preference
+	if ( hasReceivedRemotePreferences( state ) ) {
+		const recentSites = getPreference( state, 'recentSites' );
+		if ( selectedSite.ID !== recentSites[ 0 ] ) {
+			context.store.dispatch( savePreference( 'recentSites', uniq( [
+				selectedSite.ID,
+				...recentSites
+			] ).slice( 0, 3 ) ) );
```

Comment:
>Also, do you think we should so much preprocessing before dispatch?

It might be possible to move this into a selector which merges the currently selected site ID† into the recent sites preference array. We'd still need to do a comparison at some point to check that we're not saving the preference if nothing has changed.

† This relies on the action having set the selected site ID occurring synchronously... which is true, but makes me feel a little uncomfortable (e.g. if we ever want to consider something like https://github.com/chikeichan/redux-worker )